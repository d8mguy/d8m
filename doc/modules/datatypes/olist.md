## Olist

Module olist provides ordered lists, i.e., lists stored in ascending order on `$<`. These are fairly different from regular lists, and the methods are different from those of list. In fact, olists as done here are really implementing sets, and the module includes an STMap statement so this can happen automatically. The correctness of this depends on assumptions discussed in the next few paragraphs.

There are two typegen definitions here: of `olist` and `olist0`. `Olist0` is for ordered types (types with a `$<` method); it
has a different identifier because typegens don't overload in d8m. The `olist` typegen has 3 arguments: the element type `T`, an ordered type that elements can be extracted to, and a key extraction function `oextract`. Thus, for `olist`, the type `T` is not ordered but `oextract` finds something within it that is. In more standard database language, oextract provides the _key_. 

Now let's talk about assumptions made. First, in d8m we always assume that the `$<` symbol designates a total order (i.e., not a partial order). This is defined to mean that every pair of elements is comparable, that is, for each pair of elements `t1`, `t2` we have `t1 < t2 || t1 == t2 || t2 < t1`. 

Second, in `olist` specifically, we assume that `oextract` is injective when restricted to the domain of any given olist. A function `f` is injective just in case for any `t1`, `t2`, `f(t1)` `==` `f(t2)` implies `t1` `==` `t2`.  

In practice, `olist0` will be used for totally ordered base types like numbers and strings, plus tuple types provided with a `$<` method, while `olist` will be used for tuple types where we don't want to fix the ordering into the type itself, but the tuple type contains one or more ordered attributes that can serve as keys. Consider `accts:list(T)` where `T` has various attributes, including 

    attribute email:string
Now imagine defining `oextract` as

    \(x:T) { x.email }
In database language, we're saying the `email` attribute is the key. And in that mindset, we know that when adding items to `accts`, we need to ensure that the `email` attribute is never duplicated. 

This is precisely what it means in mathematical language to say that `oextract` is injective, when restricted to `accts`. As a tuple type, it's easy to create two distinct items with the same `email` attribute. That means that as a mathematical function, `oextract` is not injective. It's in the nature of tuple types that it cannot be. So the "restricted to `accts`" clause is important here. It means that although the tuple type `T` can easily have distinct elements with the same value of `email`, the collection of `T`'s in `accts` will not. 

It should be easy to see why this condition isn't expressible within d8m's type language (and arguably, isn't a proper thing for type systems to attempt to express at all). D8m's type system is about the structure of entities; this condition concerns constraints on their contents. 

Making this assumption ensures that there cannot be consecutive elements in the olist that map to the same key but are distinct, which assures us that if we're looking for some `x:T` in `accts` and find one whose `oextract` matches at index `k`, then `x` is in `accts` iff `accts[k] == x`. If in addition, we insert into the list only elements that aren't already in it, then we know the list is a set. `Olist` provides an `insertIfUniq` method to make this easy, and generally assumes you don't insert duplicates. However, it doesn't enforce this (for example, by checking for duplication in the `insert` method) for two reasons. First, it's quite common in practice to have code structured so that duplicate insertion is impossible; in this case, embedding a check in `insert` loses performance for no reason. Second, if a duplication check in `insert` fails, the correct response depends on the context &mdash; it might be to panic, to return silently, or return some kind of error message. The solution I've adopted is to explain all this in detail and then make both methods available: `insert` for when the code structure makes duplication impossible; `insertIfUniq` for when a duplicate should be ignored. 

An alternative design choice would be to drop the "injective when restricted" condition. Doing this would weaken one of the really attractive things about olists, which is that the worst case lookup time is `O(lg(n))`. Since you're not guaranteed that the element returned by binary search is the element you're looking for, you must follow each binary search with a linear one that continues until the key changes in the olist. In general, this linear search can be arbitrarily long, and unless you cache the keys, you may be calling `oextract` a lot as well. Not only does this choice reduce the attractiveness of olist as a data representation, it adds very little, since the cases where olist makes sense pretty much always have a workable key. 

#### type olist

    [olist(T, U, oextract): elts]

where `elts: list(T)`.

    val olist = \typegen(T::entity, U::ordered, oextract:\(T)->U) {
        val OLT = olist(T, U, oextract)
        method index = \(oelt: U) -> integer {...}
        method insert = \mod(elt: T) -> list(T) {...}
        method insertIfUniq = \mod(elt: T) -> nothing {...}
        method count = \() -> integer {...}
    	method remove = \mod(elt: T) -> nothing {...}
        method $in = \(elt: T) -> boolean {...}
        method keyIn = \(oelt: U) -> boolean {...}
        method get = \(oelt: U) -> nilPossible(T) {...}
        method getOk = \(oelt: U) -> T {...}
        method before = \(elt: T) -> nilPossible(T) {...}
        method after = \(elt: T) -> nilPossible(T) {...}
        method rvalindex = \(inx: integer) -> T {...}
        method diff = \(other: OLT) -> OLT {...}
        method lithook4Set = \mod(elts: list(T)) -> OLT {...}
        method popb = \mod() -> nilPossible(T)
        method popf = \mod() -> nilPossible(T)
        method minelt = \() -> nilPossible(T)
        method maxelt = \() -> nilPossible(T)
    }

Method `index` provides direct access to what binary search returns. But use it carefully since any insert or remove can change the index.

Method `remove` removes the element, does nothing if element isn't present.

Method `$in` tells if the given elt is in the olist  (only correct if $< is a total order).

Method `keyIn` is like $in but only tests the key.

Method `get` returns the (first) elt whose oextract is oelt, or nil.

Method `getOk` is as get except you know the match will succeed so the return type isn't nilpossible.

Method `before` returns the greatest list elt <= T or nil if either list is empty or T is < the first elt of the list.

Method `after` returns the least list elt >= T or nil if either list is empty or T is > the last elt of the list.

Method `insertIfUniq` returns "unless(elt in self) insert(elt)", hand written because compiler doesn't do it well enough yet.

Method `rvalindex` can be helpful for types that use olists.

Method `diff` produces self - other, ie elts in self and not in other, in the manner of the set operation.

Method `lithook4Set` is defined to support stmapping from set to olist.

Method `popf` removes first element. (Note: nilPossible.)

Method `popb` removes last element. (Note: nilPossible.)

Method `minelt` returns first element. (Note: nilPossible.)

Method `maxelt` returns last element. (Note: nilPossible.)

#### type olist0

    val olist0 = \typegen(T0::ordered) {...}

`Olist0` simply instantiates `olist`:

    val olist0 = \typegen(T0::ordered) { olist(T0, T0, \(t: T0) {t}) }
