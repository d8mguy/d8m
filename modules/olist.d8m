// Module olist provides ordered lists, ie lists stored in ascending order on $<; methods are different from those of list.
// It also defines an stmapping for olists to implement sets; the correctness of this depends on assumptions discussed
// in the next few paragraphs.
//
// There are two typegen definitions here: of olist and olist0. Olist0 is for ordered types (types with a $< method); it
// uses two identifiers because typegens don't overload in d8m. The olist typegen has 3 arguments: the element type `T`, an
// ordered type that elements can be extracted to, and a key extraction function `oextract`. The code here assumes that
// `oextract` totally orders `T`. This is subtle and isn't expressible in d8m's type scheme (or any type scheme): it means that
// the order induced by oextract is a total order in the context of the list. D8m's interpretation of the
// typepred `ordered` is discussed a bit in the spec, d8m infrastructure (if not the language itself) is built on the
// assumption that methods named `$<` designate total orders. This means (by definition) that
// for each pair of elements `t1`, `t2` we have `t1 < t2 || t1 == t2 || t2 > t1`.
//
// A concrete example may help clarify what's at stake here. Let `T` be a tuple representing accounts for some service.
// Any collection of these accounts should be a set -- no dups. If stored as an olist(T), there will normally be at least
// one attribute that's unique to each `T` -- an email address, account number, perhaps a creation
// time. There may also be attributes whose values can be duplicated among entities -- date of birth, number of months a
// subscriber, estimated net worth. All these examples are strings or numbers, so they have types that are totally ordered.
// But in the collection, some of them are keys (in the database sense) and some aren't. Your extract function should return
// a key, or if your collection doesn't have any single attributes that are keys it may be keyable with multiple attributes.

// In an olist(T) there will be some
// attribute (possibly more than one) that's unique to each entity -- an email address, account number, perhaps a creation
// time. There may also be attributes whose values can be duplicated among entities -- date of birth, number of months a
// subscriber, estimated net worth. Both the unique and non-unique things may be ordered. However, an extraction function that
// returns the date of birth as an ordered type will _not_ turn T into an ordered type in the accounts list.
// When $< is defined on numbers or strings, it satisfies the total ordering predicate by definition but when we create a
// collection of tuples, the truth of the total ordering predicate for any given ordered attribute depends on the collection. This
// truth is reflected in the fact that when we create a new account, we check whether the email address is already used in
// some existing account, or ther fact that if we use an account number for indexing the accounts, we take care to always
// allocate new accounts from unused numbers.
//
// It's worth noting that olists are still lists in the sense that they allow duplicates, what's forbidden is duplications
// of the oextract value on distinct elements. However, as a practical matter, olists tend to be most useful when used like
// sets, i.e. without duplicates. This module supports this pattern in a few ways: insertIfUnique preserves a non-duplicates property,
// and the lithook4Set method uses this one to construct a list without duplicates. Also, it defines the stmapping for olists as sets.
//
// We take this approach because the use case for non-total orders is limited, and coding for the more general case is
// somewhat slower. Note that if you want some collection in an ordering that isn't total (such as on birth year), sort
// works fine.
val olist = \typegen(T::entity, U::ordered, oextract:\(T)->U) {
	val OLT = extend tuple(private x: list(T)) where {
		assert enumerable(x)
		// this is the core binsearch algo; it's private but exported via index.
		// returns index of greatest elt <= oelt or -1 if no such elt
		private method indexA = \imp(oelt: U) {
			if(x == [] || oextract(x.last) < oelt) { -1 }
			else {
				var mn = 0, mx = x.count-1, cur = mx/2
				while(mn <= mx) {
					val ocur = oextract(x[cur])
					if(ocur < oelt) mn = cur+1
					else if(ocur == oelt) break
					else mx = cur-1
					cur = mn + (mx - mn)/2
				}
				cur
			}
		}

		// provides direct access to what binary search returns. But use it carefully since any insert or remove
		// can change the index.
		method index = \(oelt: U) { indexA(oelt) }
		method insert = \mod(elt: T) {
			val inx = indexA(oextract(elt))
			if(inx < 0) x.pushb(elt)
			else x.insert(inx, elt)
		}
		// returns "unless(elt in self) insert(elt)", hand written because compiler doesn't do it well enough yet
		method insertIfUniq = \mod(elt: T) {
			val inx = indexA(oextract(elt))
			if(inx < 0) x.pushb(elt)
			else if(x[inx] != elt) x.insert(inx, elt)
		}
		// Standard constructor: allow dups
		method lithook = \mod(elts: list(T)) {
			each(e^elts) self.insert(e)
		}
		// Non-standard constructor: can restrict dups. This is a bit nonsensical, it only makes sense with noDups set to true.
		// But optional args don't work in method decls yet so this is a way to get a TC-distinct lithook.
		method lithook = \mod(elts: list(T), noDups: boolean) {
			if(noDups) each(e^elts) self.insertIfUniq(e)
			else each(e^elts) self.insert(e)
		}
		method count = \() { x.count }
		// removes the element, does nothing if element isn't present.
		method remove = \mod(elt: T) {
			val inx = indexA(oextract(elt))
			if(inx >= 0 && x[inx] == elt) x.removeIndex(inx)
		}
		// Tells if the given elt is in the olist (only correct if $< is a total order)
		method $in = \(elt: T) -> boolean {
			val inx = indexA(oextract(elt))
			inx >= 0 && x[inx] == elt
		}
		// Is like $in but only tests the key
		method keyIn = \(oelt: U) -> boolean {
		    val inx = indexA(oelt)
		    inx >= 0 && oextract(x[inx]) == oelt
		}
		// Returns the (first) elt whose oextract is oelt, or nil
		method get = \(oelt: U) -> nilPossible(T) {
		    val inx = indexA(oelt)
		    if(inx >= 0 && oextract(x[inx]) == oelt) x[inx] else nil
		}
		// Is same as get except you know the match will succeed
		method getOK = \(oelt: U) { always(aa~get(oelt), aa != nil) }

		// returns the greatest list elt <= T or nil if either list is empty or T is < the first elt of the list.
		method before = \(elt: T) {
		    val oelt = oextract(elt)
		    if(x.count == 0 || oelt < oextract(x.head)) nil
		    else if(oextract(x.last) < oelt) x.last
		    else { val einx = indexA(oelt); if(x[einx] == oelt) x[einx] else if(einx > 0) x[einx-1] else nil }
		}
		// returns the least list elt >= T or nil if either list is empty or T is > the last elt of the list.
		method after = \(elt: T) {
		    val oelt = oextract(elt)
		    if(x.count == 0 || oextract(x.last) < oelt) nil
		    else x[indexA(oelt)]
		}
		// can be helpful for types that use olists; any mod operation invalidates the value retrieved here.
		method rvalindex = \(inx: integer) { x[inx] }

		// produces self - other, ie elts in self and not in other, in the manner of the set operation.
		method diff = \imp(other: OLT) {
            var rsltX: list(T) = []     // rsltX will be ordered but not declared as such
            if(x.count > other.count) {
                // here, delete elements of other from a copy of self
                rsltX = copy(x)
                if(other.count > 0) {
                    var inx = indexA(oextract(other[0]))        // inx is indexA in self of other[oinx]
                    var oinx = 0
                    if(inx >= 0) {
                        loop {
                            if(other[oinx] == rsltX[inx]) rsltX.removeIndex(inx)
                            oinx += 1
                            if(oinx >= other.count) break
                            while(oextract(rsltX[inx]) < oextract(other[oinx]) && inx < rsltX.count) inx += 1
                            if(inx == rsltX.count) break
                        }
                    }
                }
            } else {
                // here, x.count <= other.count so build up from empty
        		var inx = 0
                if(other.count > 0) {
                    var xoInx = indexA(oextract(other[0]))
                    var oinx = 0
                    if(xoInx < 0 || xoInx == x.count) rsltX = copy(x)       // no ovlp, so no need to scan
                    else {
                        rsltX.append(x[inx...xoInx])
                        inx = xoInx
                        loop {
                            val key1 = oextract(x[inx])
                            while(oinx < other.count && oextract(other[oinx]) < key1) oinx += 1
                            if(oinx < other.count && x[inx] != other[oinx]) rsltX.pushb(x[inx])
                            inx += 1
                            if(inx >= x.count) break
                            val key2 = oextract(x[inx])
                            while(oinx < other.count && oextract(other[oinx]) < key2) oinx += 1
                            if(oinx >= other.count) break
                        }
                    }
                }
            }
            var rslt: OLT = []
            rslt.x = rsltX          // smash
            rslt
		}

		// contains(self, other) is true just in case every elt of other is in self. IOW, "other subset self" for sets.
		method contains = \imp(other: OLT) {
		    if(other.count == 0) return true
            var ocur = other[0]
            var oinx = 1
            var sinx = indexA(oextract(ocur))
            loop {
                if(sinx < 0 || sinx >= self.count || x[sinx] != ocur) return false
                if(oinx == other.count) break
                ocur = other[oinx]
                oinx += 1
                sinx += 1
                while(sinx < x.count && x[sinx] != ocur) {
                    if(oextract(ocur) <= oextract(x[sinx])) return false
                    sinx += 1
                }
            }
            true
		}

		// removes last element which will be max according to the ordering. Note that this is nilPossible
		method popb = \mod() { x.popb }

		// removes first element which will be min according to the ordering. Note that this is nilPossible
		method popf = \mod() { x.popf }

		// first element is min according to the ordering. Note that this is nilPossible
		method minelt = \() -> nilPossible(T) { x.head }

		// last element is max according to the ordering. Note that this is nilPossible
		method maxelt = \() -> nilPossible(T) { x.last }
	}
	OLT
}

// Olist0 simply instantiates olist.
val olist0 = \typegen(T0::ordered) { olist(T0, T0, \(t: T0) {t}) }

// The following causes the given STMap to be defined as soon as olist and set are both in the bdgenv.
// Note that the clever (and O(N)) version of equality is not mapped here, so a set mapped to olist will use
// the naive (and O(N^2)) equality checking algo in the defn of set. Why? The issue has to do with whether the
// oextract fn is injective or not. If it is, then the ordering of the olist is total. If not, then elts equiv 
// under oextract can be in any order in the olist so equality checking has to find such equiv elts (which will 
// be adjacent in the olist) and check them for equality with the naive algo. The clever algo is elegantly written:
//     method eq = \(other:set(T0)) { other.count = self.count && all([s~x, o~other].xp.{s = o}) }
// Obviously, the version with proper equiv class checking can be written. In my opinion, the clearest way to 
// write it is imperative and it'll definitely be longer and harder to understand than the above. 
// Eventually, when manifest works as specified, I could define the equality predicate condition on injectiveness 
// of oextract. That's the proper solution. 
oncondition defined(set, type) {
    given(T0::entity, OT::ordered, fn:\(T0)->OT, y: T0, yl: list(T0), ys: set(T0)) {
        STMap {
            source = set(T0)
            target = olist(T0, OT, fn)
            direct count, $in
            self.lithook(yl) => lithook(self, yl, true)
            self.add(y) => self.insertIfUniq(y)
            self.remove(y) => self.remove(y)
            self.subset(ys) => self.contains(cast(ys, target))
            enumerable(self) => self
        }
    }
    given(OT::ordered, y:OT, yl: list(OT), ys: set(OT)) {
        STMap {
            source = set(OT)
            target = olist0(OT)
            direct count, $in
            self.lithook(yl) => lithook(self, yl, true)
            self.add(y) => self.insertIfUniq(y)
            self.remove(y) => self.remove(y)
            self.subset(ys) => self.contains(cast(ys, target))
            enumerable(self) => self
        }
    }
}
export olist, olist0
