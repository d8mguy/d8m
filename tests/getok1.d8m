// There was an inlining bug in getOK; this test checks that it's fixed.
// Note that there's no theory reason why getOK works here, it just happens to, given what rand.Intn does with std seed.
import "olist" melted
import go "math/rand"
val tt0 = tuple(x, y: integer)
val olt = olist(tt0, integer, \(t: tt0) { t.y })
val foo = \imp() {
    val randx = rand.New(rand.NewSource(1))
    val alist = (1..20).{ [tt0: randx.Intn(30), randx.Intn(10)+30] }
    val olist = [olt: alist]
    each(key^34..38) {
        println(olist.getOK(key))
    }
}
foo()
