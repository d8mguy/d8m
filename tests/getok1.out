[28; 34]
[0; 35]
[15; 36]
[27; 37]
[7; 38]
